src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -1489,6 +1489,211 @@ static int M1(bool? b1, bool? b2)
+                //             _ = t switch { (3, 4) => 1 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(9, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException()");
+        }
+
+        [Fact]
+        public void UnmatchedInput_03()
+        {
+            var source =
+@"using System;
+public class C
+{
+    static void Main()
+    {
+        var t = (1, 2);
+        try
+        {
+            _ = t switch { (3, 4) => 1 };
+        }
+        catch (MatchFailureException ex)
+        {
+            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex.GetType().Name);
+        }
+    }
+}
+namespace System
+{
+    public class MatchFailureException : InvalidOperationException
+    {
+        public MatchFailureException() {}
+        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public object UnmatchedValue { get; }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (9,19): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //             _ = t switch { (3, 4) => 1 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(9, 19)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException((1, 2))");
+        }
+
+        [Fact]
+        public void UnmatchedInput_04()
+        {
+            var source =
+@"using System;
+public class C
+{
+    static void Main()
+    {
+        try
+        {
+            _ = (1, 2) switch { (3, 4) => 1 };
+        }
+        catch (MatchFailureException ex)
+        {
+            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine(ex.GetType().Name);
+        }
+    }
+}
+namespace System
+{
+    public class MatchFailureException : InvalidOperationException
+    {
+        public MatchFailureException() {}
+        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
+        public object UnmatchedValue { get; }
+    }
+}
+";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (8,24): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
+                //             _ = (1, 2) switch { (3, 4) => 1 };
+                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(8, 24)
+                );
+            CompileAndVerify(compilation, expectedOutput: "MatchFailureException()");
```

Comment:
>CompileAndVerify(compilation, expectedOutput: "MatchFailureException()"); [](start = 12, length = 73)

Based on the latest LDM decision, I would expect the value to be available for this scenario.

Reply:
> The generated code does not construct a tuple object for the switch expression, so that tuple object cannot be trivially boxed into object

I got completely different impression specifically about tuples for the LDM discussion. It would be good to bring this specific scenario to LDM for confirmation. This issue doesn't block this PR though.

---
In reply to: [234384692](https://github.com/dotnet/roslyn/pull/31220#discussion_r234384692) [](ancestors = 234384692)